src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultUsernamePasswordCredentialImpl.java

```diff
@@ -46,6 +48,14 @@ public VaultUsernamePasswordCredentialImpl(CredentialsScope scope, String id,
         password = null;
     }
 
+    private Object writeReplace() {
+        if (Channel.current() == null) {
+            return this;
+        }
+        VaultUsernamePasswordCredentialsSnapshotTaker snapshotter = new VaultUsernamePasswordCredentialsSnapshotTaker();
+        return snapshotter.snapshot(this);
+    }
```

Comment:
this is wrong, you should not be calling snapshot anywhere in this code (unless you are actually *using* credentials.

Rather it is the consumer that should snapshot them, the provider just provides the CredentialsSnapshotTaker implementation.